# Question 1: Install Spark and PySpark

    Install Spark
    Run PySpark
    Create a local spark session
    Execute spark.version.

What's the output?

In [2]:
import pyspark
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
25/07/06 20:37:39 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [3]:
spark.version

'4.0.0'

Answer: 4.0.0

# Question 2: Yellow October 2024

Read the October 2024 Yellow into a Spark Dataframe.

Repartition the Dataframe to 4 partitions and save it to parquet.

What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.

    6MB
    25MB
    75MB
    100MB


In [4]:
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2024-10.parquet

--2025-07-06 20:40:03--  https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2024-10.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 2600:9000:234c:fc00:b:20a5:b140:21, 2600:9000:234c:9200:b:20a5:b140:21, 2600:9000:234c:3800:b:20a5:b140:21, ...
Connecting to d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)|2600:9000:234c:fc00:b:20a5:b140:21|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 64346071 (61M) [binary/octet-stream]
Saving to: ‘yellow_tripdata_2024-10.parquet’

yellow_tripdata_202 100%[===================>]  61.36M  3.18MB/s    in 21s     

2025-07-06 20:40:25 (2.99 MB/s) - ‘yellow_tripdata_2024-10.parquet’ saved [64346071/64346071]



In [10]:
df_yellow_202410 = spark.read \
    .option('header', 'true') \
    .parquet('yellow_tripdata_2024-10.parquet')

In [11]:
df_yellow_202410.printSchema()

root
 |-- VendorID: integer (nullable = true)
 |-- tpep_pickup_datetime: timestamp_ntz (nullable = true)
 |-- tpep_dropoff_datetime: timestamp_ntz (nullable = true)
 |-- passenger_count: long (nullable = true)
 |-- trip_distance: double (nullable = true)
 |-- RatecodeID: long (nullable = true)
 |-- store_and_fwd_flag: string (nullable = true)
 |-- PULocationID: integer (nullable = true)
 |-- DOLocationID: integer (nullable = true)
 |-- payment_type: long (nullable = true)
 |-- fare_amount: double (nullable = true)
 |-- extra: double (nullable = true)
 |-- mta_tax: double (nullable = true)
 |-- tip_amount: double (nullable = true)
 |-- tolls_amount: double (nullable = true)
 |-- improvement_surcharge: double (nullable = true)
 |-- total_amount: double (nullable = true)
 |-- congestion_surcharge: double (nullable = true)
 |-- Airport_fee: double (nullable = true)



In [12]:
df_yellow_202410.repartition(4) \
    .write.parquet('data/pq/yellow/2024/10/')

In [13]:
ls -lah data/pq/yellow/2024/10

total 190472
-rw-r--r--@  1 romadv-mac  staff     0B  6 Jul 20:47 _SUCCESS
drwxr-xr-x@ 12 romadv-mac  staff   384B  6 Jul 20:47 ./
-rw-r--r--@  1 romadv-mac  staff     8B  6 Jul 20:47 ._SUCCESS.crc
drwxr-xr-x@  3 romadv-mac  staff    96B  6 Jul 20:47 ../
-rw-r--r--@  1 romadv-mac  staff   179K  6 Jul 20:47 .part-00000-6d5c2041-dbcb-4e63-8f20-1a3d09b3c245-c000.snappy.parquet.crc
-rw-r--r--@  1 romadv-mac  staff   179K  6 Jul 20:47 .part-00001-6d5c2041-dbcb-4e63-8f20-1a3d09b3c245-c000.snappy.parquet.crc
-rw-r--r--@  1 romadv-mac  staff   179K  6 Jul 20:47 .part-00002-6d5c2041-dbcb-4e63-8f20-1a3d09b3c245-c000.snappy.parquet.crc
-rw-r--r--@  1 romadv-mac  staff   179K  6 Jul 20:47 .part-00003-6d5c2041-dbcb-4e63-8f20-1a3d09b3c245-c000.snappy.parquet.crc
-rw-r--r--@  1 romadv-mac  staff    22M  6 Jul 20:47 part-00000-6d5c2041-dbcb-4e63-8f20-1a3d09b3c245-c000.snappy.parquet
-rw-r--r--@  1 romadv-mac  staff    22M  6 Jul 20:47 part-00001-6d5c2041-dbcb-4e63-8f20-1a3d09b3c245-c000.snappy.parquet

Answer: 25MB

# Question 3: Count records

How many taxi trips were there on the 15th of October?

Consider only trips that started on the 15th of October.

    85,567
    105,567
    125,567
    145,567


In [18]:
df_yellow_202410.show(5)

+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+
|       2| 2024-10-01 00:30:44|  2024-10-01 00:48:26|              1|          3.0|         1|                 N|         162|         246|           1|       18.4|  1.0|    0.5|       1.

In [15]:
df_yellow_202410.createOrReplaceTempView('trips_data')

In [17]:
spark.sql("""
SELECT COUNT(*)
FROM trips_data
WHERE DATE(tpep_pickup_datetime) = '2024-10-15'
""")

+--------+
|count(1)|
+--------+
|  128893|
+--------+



Answer: 125,567

# Question 4: Longest trip

What is the length of the longest trip in the dataset in hours?

    122
    142
    162
    182


In [23]:
spark.sql("""
WITH t1 AS (
SELECT DATE_DIFF(tpep_dropoff_datetime, tpep_pickup_datetime) as trip_days
FROM trips_data
)
SELECT MAX(trip_days) * 24
FROM t1
""").show()

+---------------------+
|(max(trip_days) * 24)|
+---------------------+
|                  168|
+---------------------+



Answer: 162

# Question 5: User Interface

Spark’s User Interface which shows the application's dashboard runs on which local port?

    80
    443
    4040
    8080


Answer: 4040

# Question 6: Least frequent pickup location zone

Load the zone lookup data into a temp view in Spark:

wget https://d37ci6vzurychx.cloudfront.net/misc/taxi_zone_lookup.csv

Using the zone lookup data and the Yellow October 2024 data, what is the name of the LEAST frequent pickup location Zone?

    Governor's Island/Ellis Island/Liberty Island
    Arden Heights
    Rikers Island
    Jamaica Bay


In [24]:
df_zones = spark.read \
    .option("header", "true") \
    .csv('data/lookup/taxi_zone_lookup.csv')

In [25]:
df_zones.show(5)

+----------+-------------+--------------------+------------+
|locationid|      borough|                zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
+----------+-------------+--------------------+------------+
only showing top 5 rows


In [26]:
df_zones.createOrReplaceTempView('zones')

In [30]:
spark.sql("""
WITH least AS (
    SELECT t.PULocationID
    , COUNT(*) AS num_pickups
    FROM trips_data t
    GROUP BY t.PULocationID
    ORDER BY num_pickups ASC
    LIMIT 1
)
SELECT zone, num_pickups
FROM least l
INNER JOIN zones z ON l.PULocationID = z.locationid
""").show()

+--------------------+-----------+
|                zone|num_pickups|
+--------------------+-----------+
|Governor's Island...|          1|
+--------------------+-----------+



Answer: Governor's Island/Ellis Island/Liberty Island

In [31]:
spark.stop()